lib/pricing-poller.js

```diff
@@ -0,0 +1,242 @@
+const log = require('./log');
+const assert = require('assert');
+const Iterate = require('taskcluster-lib-iterate');
+
+function toFloat(v) {
+  if (typeof v === 'number') {
+    return v;
+  } else if (typeof v === 'string') {
+    return parseFloat(v, 10);    
+  } else {
+    throw new Error('Number is not and cannot be a float');
+  }
+}
+
+class PricingPoller {
+
+  constructor({ec2, timePeriod, regions, runaws, pollDelay, monitor}) {
+    assert(typeof ec2 === 'object');
+    this.ec2 = ec2;
+
+    assert(typeof runaws === 'function');
+    this.runaws = runaws;
+
+    assert(typeof regions === 'object');
+    assert(Array.isArray(regions));
+    this.regions = regions;
+
+    // Number of minutes to consider for pricing
+    assert(typeof timePeriod === 'number');
+    this.timePeriod = timePeriod;
+
+    // Number of seconds to wait before polling for pricing again
+    assert(typeof pollDelay === 'number');
+    this.pollDelay = pollDelay;
+
+    // Monitor instance
+    assert(typeof monitor === 'object');
+    this.monitor = monitor;
+
+    this.iterator = new Iterate({
+      maxIterationTime: 1000 * 60 * 15,
+      watchDog: 1000 * 60 * 15,
+      maxFailures: 10,
+      waitTime: pollDelay,
+      handler: async (watchdog) => {
+        try {
+          await this.poll();
+        } catch (err) {
+          console.dir(err.stack || err);
+          monitor.reportError(err, 'Error polling pricing data');
+        }
+      },
+    });
+
+    // It is important for getPrices() behaviour that this variable be set to
+    // undefined before prices are found so we can throw an error appropriately
+    this.prices = undefined;
+  }
+
+  start() {
+    return this.iterator.start();
+  }
+
+  stop() {
+    return this.iterator.stop();
+  }
+
+  /**
+   * Take a list of {price, time} and return a price that we should use
+   */
+  _findSpotPriceForInstanceType(series) {
+    let now = new Date();
+    let maxPrice = 0;
+    for (let {price, time} of series) {
```

Comment:
Are you in the anti lodash camp? If not, we can replace this loop with `_.maxBy(series, 'price')` maybe?